### PR TITLE
[clang][include-tree] Fix spurious dependencies

### PR DESCRIPTION
**Pull request by jansvoboda11**
This patch fixes an incorrect assumption that only spurious imports cause both `IncludeTreeBuilder::moduleImport()` and `IncludeTreeBuilder::exitedInclude()` to be called for the same directive. This can also happen for `#include` of a header that belongs to the module being implemented that previously got transitively loaded from a PCM.

Instead of trying to guess that a submodule was a spurious import (by checking that it was not imported, is not from module file but its parent is from module file), we just mark it missing from umbrella header when we detect this. This allows us to stop putting the temporary `ModuleImport` onto the stack in one function and later popping and replacing it by `SpuriousImport` in another function.

rdar://123253616

Cherry-pick of apple/llvm-project/pull/8264

### CODE REVIEW

**Comment by jansvoboda11** [comment]
@swift-ci please test

File: clang/lib/Tooling/DependencyScanning/IncludeTreeActionController.cpp [hunk 1]
```diff
@@ -457,20 +452,25 @@ void IncludeTreeBuilder::exitedInclude(Preprocessor &PP, FileID IncludedBy,
   SourceManager &SM = PP.getSourceManager();
   std::pair<FileID, unsigned> LocInfo = SM.getDecomposedExpansionLoc(ExitLoc);
 
-  // If the file includes already has a node for the current source location, it
-  // must be an import that turned out to be spurious.
-  auto &CurIncludes = IncludeStack.back().Includes;
```
Comment (benlangmuir) [inline clang/lib/Tooling/DependencyScanning/IncludeTreeActionController.cpp:462]: Nit: would be nice to keep a short comment here about what's going on; the code itself is understandable but a bit dense to read through just to see that it's about spurious imports.

**Review by benlangmuir — APPROVED** [review]

**Comment by jansvoboda11** [comment]
https://github.com/apple/swift/pull/71886
@swift-ci please test Windows Platform

**Comment by jansvoboda11** [comment]
https://github.com/apple/swift/pull/71886
@swift-ci please test

**Comment by jansvoboda11** [comment]
@swift-ci please test

**Pull request merged**